### DISCUSSION TITLE
Properly handle leading and trailing spaces in tag names and values

### DISCUSSION THREAD

**Discussion by jonathanwcrane** [cloud-custodian/cloud-custodian#7314] [General]
Currently, cloud custodian considers leading and trailing spaces as deviations from the approved tags and values.  This is obviously not what any user intends when typing in a tag or value when creating a resource.  So cloud custodian would ideally:

- ignore leading and trailing spaces
or, better yet
- actively fix all instances of leading or trailing spaces in tag names and values.

**Comment by kapilt** [comment]
use value_type: normalize

**Comment by jonathanwcrane** [comment]
That's only for the value, right?  It's just as easy to accidentally put a trailing space after the tag name; is that taken care of as well?

**Comment by jonathanwcrane** [comment]
So I'm digging into the code and rename_tag in tags.py might be a candidate to use to fix the *tag values* themselves.  Is there anyone familiar with the code that can answer questions like to what does the w in `w.submit()` refer on line 658 of tags.py?

**Comment by kapilt** [comment]
thats concurrent.futures executor syntax. see https://docs.python.org/3/library/concurrent.futures.html

**Comment by kapilt** [comment]
it would be nice to support this perhaps a new dedicated filter, i'm cautious about overloading the extant value filter for it as out of scope, but do see a need for this sort of normalization across the tag key space.

**Comment by jonathanwcrane** [comment]
Seems to make sense.  I'm just trying to think of any valid reason someone would have for a leading or trailing space in a tag name or value.

Now that I think about it, I think AWS should be filtering it at creation/update time themselves!
